doc/subsystems/shell/shell.rst

```diff
@@ -0,0 +1,296 @@
+  and color printing.
+* Support for multiline commands.
+* Built-in handler to display help for the commands.
+* Support for wildcards: ``*`` and ``?``.
+* Support for meta keys.
+
+The module can be connected to any transport for command input and output.
+At this point, the following transport layers are implemented:
+
+* UART
+
+See the :ref:`shell_api` documentation for more information.
+
+Usage
+*****
+
+Use the :c:macro:`SHELL_DEFINE` macro to create an instance of the shell.
+Pass the expected newline character to this macro: either ``\r`` or ``\n``,
+otherwise the shell will not respond correctly to the :kbd:`Enter` key.
+
+The following code shows a simple use case of this library:
+
+.. code-block:: c
+
+	/* Defining shell backend */
+	SHELL_UART_DEFINE(shell_transport_uart);
+
+	/* Creating shell instance */
+	SHELL_DEFINE(uart_shell, "uart:~$ ", &shell_transport_uart, '\r', 10);
+
+	void main(void)
+	{
+		(void)shell_init(&uart_shell, NULL, true, true, LOG_LEVEL_INF);
+	}
+
+	/* Creating subcommands (level 1 command) array for command "demo".
+	 * Subcommands must be added in alphabetical order
+	 */
+	SHELL_CREATE_STATIC_SUBCMD_SET(sub_demo)
+	{
+		/* Alphabetically sorted. */
+		SHELL_CMD(params, NULL, "Print params command.",
+						       cmd_demo_params),
+		SHELL_CMD(ping,   NULL, "Ping command.", cmd_demo_ping),
+		SHELL_SUBCMD_SET_END /* Array terminated. */
+	};
+	/* Creating root (level 0) command "demo" */
+	SHELL_CMD_REGISTER(demo, &sub_demo, "Demo commands", NULL);
+
+	/* Creating root (level 0) command "version" */
+	SHELL_CMD_REGISTER(version, NULL, "Show kernel version", cmd_version);
+
+
+Users may use the :kbd:`Tab` key to complete a command/subcommand or to see the
+available subcommands for the currently entered command level.
+For example, when the cursor is positioned at the beginning of the command
+line and the :kbd:`Tab` key is pressed, the user will see all root (level 0)
+commands:
+
+.. code-block:: none
+
+	  clear  demo  shell  history  log  resize  version
+
+
+.. note::
+	To view the subcommands that are available for a specific command, you
+	must first type a :kbd:`space` after this command and then hit
+	:kbd:`Tab`.
+
+These commands are registered by various modules, for example:
+
+* :command:`clear`, :command:`shell`, :command:`history`, and :command:`resize`
+  are built-in commands which have been registered by
+  :file:`subsys/shell/shell.c`
+* :command:`demo` and :command:`version` have been registered in example code
+  above by main.c
+* :command:`log` has been registered by :file:`subsys/logging/log_cmds.c`
+
+Then, if a user types a :command:`demo` command and presses the :kbd:`Tab` key,
+the shell will only print the subcommands registered for this command:
+
+.. code-block:: none
+
+	  params  ping
+
+
+Commands
+********
+
+Shell commands are organized in a tree structure and grouped into the following
+types:
+
+* Root command (level 0): Created in the section variable area.
```

Comment:
`Gathered and alphabetically sorted in a dedicated memory section` i think that 'memory section' is better than 'section variable area'.